src/compiler/checker.ts

```diff
@@ -20371,10 +20375,26 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             error(currentNode, Diagnostics.Type_instantiation_is_excessively_deep_and_possibly_infinite);
             return errorType;
         }
+        const index = findActiveMapper(mapper);
+        if (index === -1) {
+            pushActiveMapper(mapper);
+        }
+        const key = type.id + getAliasId(aliasSymbol, aliasTypeArguments);
+        const mapperCache = activeTypeMappersCaches[index !== -1 ? index : activeTypeMappersCount - 1];
+        const cached = mapperCache.get(key);
+        if (cached) {
+            return cached;
+        }

```

Comment:
Since `key` and `mapperCache` aren't used for initial occurrences of mappers, computing the key can be avoided:

```suggestion
        let key: string | undefined;
        let mapperCache: Map<string, Type> | undefined;
        if (index === -1) {
            pushActiveMapper(mapper);
        } else {
            key = type.id + getAliasId(aliasSymbol, aliasTypeArguments);
            mapperCache = activeTypeMappersCaches[index];
            const cached = mapperCache.get(key);
            if (cached) {
                return cached;
            }
        }
```

Likely also needs some updates below because of the `undefined`s.

I also wonder if it's worth making the allocation of the `Map` lazy (and/or to keep it around if it's empty)

Reply:
Or possibly only push if the type could potentially recurse (depending on the type flags), although I suspect that delaying the allocation of `Map` should be similar in perf; the push/pop operations themselves should be rather cheap.

Reply:
Thanks for the suggestions. I'm not sure if this PR is going to make it though - so, for now, I'm refraining from making any changes to it.